### DISCUSSION TITLE
unable load firmware

### DISCUSSION THREAD

**Discussion by rinrin-** [microsoft/WSL#8420] [General]
### Version

Microsoft Windows [Version 10.0.22000.651]

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

Linux version 5.10.102.1-microsoft-standard-WSL2

### Distro Version

Gentoo/Funtoo

### Repro Steps

1. run the following command in cmd:
usbipd wsl attach -b 1-X
2. after usb driver init, load another mod
# modprobe carl9170

### Expected Behavior

firmware carl9170-1.fw load successfully

### Actual Behavior

firmware carl9170-1.fw can not be loaded

### Diagnostic Logs

kernel log after enable dynamic debug:
[  719.134445] usb 1-1: new high-speed USB device number 4 using vhci_hcd
[  719.284377] usb 1-1: SetAddress Request (4) to port 0
[  719.349176] usb 1-1: New USB device found, idVendor=0846, idProduct=9001, bcdDevice= 1.06
[  719.349185] usb 1-1: New USB device strings: Mfr=16, Product=32, SerialNumber=48
[  719.349189] usb 1-1: Product: USB2.0 WLAN
[  719.349191] usb 1-1: Manufacturer: ATHER
[  719.349193] usb 1-1: SerialNumber: 12345
[  719.494320] usb 1-1: reset high-speed USB device number 4 using vhci_hcd
[  719.644405] usb 1-1: SetAddress Request (4) to port 0
[  719.839554] firmware_class: __allocate_fw_priv: fw-carl9170-1.fw fw_priv=000000009a54a0a0
[  719.839579] usb 1-1: loading /lib/firmware/updates/5.10.102.1-microsoft-standard-WSL2/carl9170-1.fw failed for no such file or directory.
[  719.839585] usb 1-1: loading /lib/firmware/updates/carl9170-1.fw failed for no such file or directory.
[  719.839590] usb 1-1: loading /lib/firmware/5.10.102.1-microsoft-standard-WSL2/carl9170-1.fw failed for no such file or directory.
[  719.839594] usb 1-1: **loading /lib/firmware/carl9170-1.fw failed for no such file or directory**.
[  719.839598] usb 1-1: Direct firmware load for carl9170-1.fw failed with error -2
[  719.839603] firmware_class: __free_fw_priv: fw-carl9170-1.fw fw_priv=000000009a54a0a0 data=0000000000000000 size=0
[  719.839608] usb 1-1: firmware not found.

but firmware is present in /lib/modules:

wsl2 ~ # ls -l /lib/firmware/carl9170-1.fw
-rw-r--r-- 1 root root 13388  4月26日 12:21 /lib/firmware/carl9170-1.fw

**Comment by rinrin-** [comment]
the firmware load code shows kernel read file from init namespace
(drivers/base/firmware_loader/main.c)
`/* load firmware files from the mount namespace of init */
rc = kernel_read_file_from_path_initns(path, fw_priv->offset,
       &buffer, msize,
       file_size_ptr,
       READING_FIRMWARE);`
I have checked with init file, it is mounted with 9p protocol:
`tools on /init type 9p (ro,relatime,dirsync,aname=tools;fmask=022,loose,access=client,msize=65536,trans=fd,rfd=7,wfd=7)`

**Reply by xunanmu** [reply]
我也遇到了，这个问题，请问你现在解决了吗？

**Comment by ghost** [comment]
Unless this is a known issue with WSL, I would suggest filing this issue with the usbipd library, which you can find [here](https://github.com/dorssel/usbipd-win/issues). Marking as external and converting to discussion.

**Reply by rinrin-** [reply]
see [comment](https://github.com/dorssel/usbipd-win/issues/390#issuecomment-1134472995) in usbipd

**Reply by xunanmu** [reply]
I'm having this problem too, I can't find the firmware path when loading the firmware
